Source/WebCore/css/CSSValueList.cpp

```diff
@@ -49,79 +100,129 @@ CSSValueList::CSSValueList(ValueSeparator separator, CSSValueListBuilder values)
 {
 }
 
-Ref<CSSValueList> CSSValueList::createCommaSeparated()
+CSSValueList::CSSValueList(ValueSeparator separator, Ref<CSSValue> value)
+    : CSSValueContainingVector(ValueListClass, separator, WTFMove(value))
 {
-    return adoptRef(*new CSSValueList(CommaSeparator));
 }
 
-Ref<CSSValueList> CSSValueList::createSpaceSeparated()
+CSSValueList::CSSValueList(ValueSeparator separator, Ref<CSSValue> value1, Ref<CSSValue> value2)
+    : CSSValueContainingVector(ValueListClass, separator, WTFMove(value1), WTFMove(value2))
 {
-    return adoptRef(*new CSSValueList(SpaceSeparator));
 }
 
-Ref<CSSValueList> CSSValueList::createSlashSeparated()
+CSSValueList::CSSValueList(ValueSeparator separator, Ref<CSSValue> value1, Ref<CSSValue> value2, Ref<CSSValue> value3)
+    : CSSValueContainingVector(ValueListClass, separator, WTFMove(value1), WTFMove(value2), WTFMove(value3))
+{
+}
+
+CSSValueList::CSSValueList(ValueSeparator separator, Ref<CSSValue> value1, Ref<CSSValue> value2, Ref<CSSValue> value3, Ref<CSSValue> value4)
+    : CSSValueContainingVector(ValueListClass, separator, WTFMove(value1), WTFMove(value2), WTFMove(value3), WTFMove(value4))
 {
-    return adoptRef(*new CSSValueList(SlashSeparator));
 }
 
 Ref<CSSValueList> CSSValueList::createCommaSeparated(CSSValueListBuilder values)
 {
     return adoptRef(*new CSSValueList(CommaSeparator, WTFMove(values)));
 }
 
-Ref<CSSValueList> CSSValueList::createSpaceSeparated(CSSValueListBuilder values)
+Ref<CSSValueList> CSSValueList::createCommaSeparated(Ref<CSSValue> value)
 {
-    return adoptRef(*new CSSValueList(SpaceSeparator, WTFMove(values)));
+    return adoptRef(*new CSSValueList(CommaSeparator, WTFMove(value)));
 }
 
 Ref<CSSValueList> CSSValueList::createSlashSeparated(CSSValueListBuilder values)
 {
     return adoptRef(*new CSSValueList(SlashSeparator, WTFMove(values)));
 }
 
-bool CSSValueContainingVector::removeAll(CSSValue& value)
+Ref<CSSValueList> CSSValueList::createSlashSeparated(Ref<CSSValue> value)
 {
-    return m_values.removeAllMatching([&value](auto& current) {
-        return current->equals(value);
-    }) > 0;
+    return adoptRef(*new CSSValueList(SlashSeparator, WTFMove(value)));
 }
 
-bool CSSValueContainingVector::removeAll(CSSValueID value)
+Ref<CSSValueList> CSSValueList::createSlashSeparated(Ref<CSSValue> value1, Ref<CSSValue> value2)
 {
-    return m_values.removeAllMatching([value](auto& current) {
-        return WebCore::isValueID(current, value);
-    }) > 0;
+    return adoptRef(*new CSSValueList(SlashSeparator, WTFMove(value1), WTFMove(value2)));
+}
+
+Ref<CSSValueList> CSSValueList::createSpaceSeparated()
+{
+    return adoptRef(*new CSSValueList(SpaceSeparator));
+}
+
+Ref<CSSValueList> CSSValueList::createSpaceSeparated(CSSValueListBuilder values)
```

Comment:
Why does this (and many other create/constructor functions) take `CSSValueListBuilder` instead of `CSSValueListBuilder&&`? Isn't it very easy for the callers to make accidental copies?

Reply:
@weinig convinced me this was the right future idiom. Basically create functions and constructors take objects, not references, and other functions take references. Maybe he can comment on it some more. It does create a risk of forgetting to move, but it comes with benefit of making it easier to construct in cases where there is not an existing object to move. I was feeling really confident about this, but now I am doubting myself

Reply:
@weinig has been suggesting this for a while but I still don't think that this is a good idea, precisely because it is too easy to forget to WTFMove() with this approach.

Reply:
With this approach, we WILL very likely forget to WTFMove() and thus end up with less efficient code, simply because we want to avoid an extra `CSSValueListBuilder { }` call at the call site whenever we don't have an r-value reference. It doesn't seem worth it.

I have r+ recently a patch adding `&&` to such parameters because the lack of consistency was bugging people too.

Reply:
OK. We can add them all; I am OK with that. I will do it in a separate patch unless something else comes up requiring me to revise this one.

Reply:
I think this depends on the primary use case. If the normal usage is to take the ownership (like here), and not doing so could be costly, then `&&` makes sense. It captures the intent of the interface in a compiler checkable manner.

Reply:
I will add in the `&&` in a future patch.